Added function for optimizing phase-shift gates

BoschSamuel opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Added the function optimize_phase_shift_gates.py in qiskit-terra/qiskit/transpiler/passes/
This is a pass for merging and canceling phase-shift gates like u1(theta), T-gates, S-gates and Z-gates and thereby optimizing the size of a DAGCircuit.

### Details and comments
Hi there,
my name is Samuel Bosch, and I am a master's student from EPFL in Switzerland. I work in the group of prof. Giovanni de Micheli on Quantum Computing. My supervisors are [dr. Mathias Soeken](https://github.com/msoeken) and [Bruno Schmitt](https://github.com/boschmitt). I started programming in Qiskit quite recently and would like to start contributing to the Qiskit library on a regular basis. This function is based on the following paper:

> Amy, Matthew, Dmitri Maslov, and Michele Mosca. "Polynomial-time T-depth optimization of Clifford+ T circuits via matroid partitioning." IEEE Transactions on Computer-Aided Design of Integrated Circuits and Systems 33.10 (2014): 1476-1489.

In addition to the functionality described in the paper, I further optimized the circuits by applying the following technique: Whenever there are several phase-shift gates, which can be merged into one phase-shift gate, there exist several possible locations within the circuit, where the new gate could be placed. The algorithms tests all possible locations, runs a cx-cancellation on the circuit for every possible location, and then measures the total number of gates for every possible location. The final location is chosen to be the one, where the circuit has the lowest number of gates. If the number of gates is the same for several possible locations, the location with the lowest circuit-depth is chosen. 

The function was tested extensively, by generating large random circuits, with many randomly chosen gates at random locations. The equivalence of the original and optimized circuits was verified by calculating and comparing the unitary operators corresponding to those two circuits.

I kindly ask you to review the code and let me know if you would like me to make any changes to it. My E-Mail address is samuel.bosch@epfl.ch

Thank you and kind regards,
Samuel Bosch

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=2229) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=2229) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=2229) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=2229) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=2229) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=2229) it.</sub>

maddy-tod commented on qiskit/transpiler/passes/optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,105 @@
+            state_tracker[i][i] = 1 # initial state
+        k = dag.width()  # number of initial variables
+        for e in dag.topological_nodes():
+            if e.data_dict['type'] == 'op':
```
This can be simplified to `e.type == 'op'`. There are a few places here where you are accessing the `data_dict` and this should not be necessary, you should be able to simply use the dot notation instead.

BoschSamuel replied:
Thank you! I changed it, and the code runs fine without accessing `data_dict ` at any place :)

maddy-tod commented on qiskit/transpiler/passes/optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,105 @@
+            else:
+                original_dag = copy.deepcopy(dag)
+                for final_gate_position in range(len(T_position_counter[key])):
+                    dag = copy.deepcopy(original_dag)
```
I think this could just be `dag = copy.deepcopy(dag)` i.e. only copying the dag once.

BoschSamuel replied:
I could be wrong, but at the moment I don't think that this would work. The `copy.deepcopy()` in line 72 is needed in every iteration, as we would otherwise loose the optimization of other phase-rotation gates made in previous runs of the `for key in T_counter:` loop. 

The `copy.deepcopy()` in line 74 is also needed in every iteration of `for final_gate_position in range(len(T_position_counter[key])):`, as the algorithm would not work otherwise. Right?

maddy-tod replied:
I meant replacing the copy on line 74 and removing the copy on line 72. It seems a copy is made, and then the only use of it is to make another copy? I think that this could been made from the original dag, but I could be mistaken! It may be more clear if the variable on line 74  is declared as something other than `dag`, perhaps something like `current_dag`

BoschSamuel replied:
Thank you for pointing this out. We had a careful look at the code, and simply removing the copy in line 72 isn't possible. Nevertheless, we declared the `dag` in line 74 as the new variable `dag_copy`, which enabled us to further simplify the code.

maddy-tod commented on qiskit/transpiler/passes/optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,105 @@
+                    phase_T_gate_equivalent = {'t': 1, 'tdg': 7, 's': 2, 'sdg': 6, 'z': 4}[e.data_dict['name']]
+                    if str(state_tracker[e.data_dict['qargs'][0][1]]) in T_counter:
+                        T_counter[str(state_tracker[e.data_dict['qargs'][0][1]])] += phase_T_gate_equivalent
+                        T_position_counter[str(state_tracker[e.data_dict['qargs'][0][1]])].append(e._node_id)
```
Instead of storing the `._node_id` is it possible to store the node itself? The `DAGCircuit` takes `DAGNodes` instead of `node_id`s now so this may clean other areas up too.

BoschSamuel replied:
Excellent idea! I will try to implement this :)

BoschSamuel replied:
I implemented the requested change. Now the nodes are stored, instead of the `._node_id`. Nevertheless, this didn't enable me to completely avoid calling `._node_id` and `.remove_op_node()`, as while optimizing the circuit, it is necessary to make a copy of a circuit, which then contains different nodes (but with the same node id).

maddy-tod commented on test/python/transpiler/test_optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,65 @@
+        backend_sim = BasicAer.get_backend('unitary_simulator')
+        result_optimized = execute(circ_new, backend_sim).result()
+        unitary_optimized = result_optimized.get_unitary(circ_new)
+        if np.sum(np.abs(unitary_optimized-unitary_original)) > 0.001:
```
Could you update your testing to use assert statements? It should just be a case of updating these end few lines, for example `assertEqual(unitary_original, unitary_optimized)`. There is a good example of how to do it [here](https://github.com/Qiskit/qiskit-terra/blob/master/test/python/transpiler/test_collect_2q_blocks.py)

BoschSamuel replied:
Thank you for the comment. We implemented everything as requested. Hopefully it is ok now.

BoschSamuel commented:
Hi there,
Sorry about the delay! I fixed the formatting and implemented the requested changes. Let me know if anything else is needed. 
Best regards, Samuel

BoschSamuel commented:
Hello again,
for some reason one of the tests starts failing, even though I just merged the master branch into phase_opt to keep it up-to-date. My code got a rating of 10.00/10, and I am unable to find a mistake on my side. Could someone please have a look, and take care of the pull request? 
Thank you!
Samuel

diego-plan9 commented:
Hi @BoschSamuel,

I just made a commit (https://github.com/Qiskit/qiskit-terra/pull/2229/commits/7465a361c250882a342937a5323179d714090e93) in the hopes of helping breaking the import cycles - in the past days there were some tweaks to the linter configuration that made it _not_ check for cycles temporarily, and we reverted that change between your latest two merge commits (that explains why it's appearing now!).

For reference, when something similar happens, you can check the logs in the CI (in the red cross - eventually you can reach [an output like this](https://travis-ci.com/Qiskit/qiskit-terra/jobs/200289592#L726), which you hopefully should also be able to reproduce by executing `make lint` on your local machine. In this case, it was mostly caused by the `import qiskit` directive - our import hierarchy is a bit convoluted, and in general a good aproach is to try to import from specific modules instead.

I'll keep an eye to see if the cycles reappear, as travis is a bit overloaded at the moment, but hopefully that will do the trick! As usual, remember to `git pull` to fetch the changes locally, as I basically committed directly to your branch for convenience.

BoschSamuel commented:
Hi @diego-plan9 ,
thank you for fixing the problem!
Best, Samuel

BoschSamuel commented:
Hey, 
it looks like some recent changes to qiskit now cause this algorithm to fail. I guess it is just a minor issue, but it would be good to know if there is an easy fix to this problem. 

Best, 
Samuel

```
    b'  File "/home/travis/build/Qiskit/qiskit-terra/qiskit/transpiler/passes/optimize_phase_shift_gates.py", line 77, in run'
    b'    circuit_length_min = len(dag.multi_graph.nodes) + 1'
    b"AttributeError: 'DAGCircuit' object has no attribute 'multi_graph'"
```

maddy-tod commented:
Hi, the multigraph has been moved to a private variable `dag._multigraph`. There is also a function `dag.node_counter()` which may do what you need.

maddy-tod commented on qiskit/transpiler/passes/optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,120 @@
+        for key in t_counter:
+            if t_counter[key] % 8 == 0:
+                for i in t_position_counter[key]:
+                    dag.remove_op_node(dag._id_to_node[i._node_id])  # Remove cancelled nodes
```
I think this can just be `dag.remove_op_node(i)`. It seems as though you are getting the id of the node, and then using this id to identify the node, but you could just use the node in the first place.

maddy-tod commented on test/python/transpiler/test_optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,70 @@
+                y = np.random.randint(n-1)
+                circ.cx(q[y+1], q[y])
+        circ_dag = circuit_to_dag(circ)
+        circ_dag = OptimizePhaseShiftGates.run(circ_dag, circ_dag)
```
It would be better to create an instance of the pass here. So something like 
```
pass = OptimizePhaseShiftGates()
pass.run(circ_dag)
```
as you shouldn't have to pass the dag to the pass twice.

1ucian0 commented on qiskit/transpiler/passes/optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,120 @@
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
```suggestion
# (C) Copyright IBM 2019.
```

1ucian0 commented on test/python/transpiler/test_optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,70 @@
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
```suggestion
# (C) Copyright IBM 2019.
```

1ucian0 commented on test/python/transpiler/test_optimize_phase_shift_gates.py:
```diff
@@ -0,0 +1,70 @@
+        i = 0
+        while i < 150:
+            i += 1
+            x = np.random.randint(9)
```
In general, we prefer deterministic tests.

BoschSamuel replied:
With "prefer" you mean I should change it now, or is that just a hint for future contributions? :)

ajavadia replied:
At some point we should move the random tests to `test/randomized` which is better suited to handle them. But until then we need them in a lot of places to increase our confidence in the code (See tests under `quantum_info` specifically). So I wouldn't remove them until they go and exist somewhere else.

BoschSamuel commented:
Hey everyone, sadly due to my current professional situation, I won't be able to look into this before February 2020.  Sorry about that! You can put this on hold until then. 

Best,
Samuel

1ucian0 commented:
Any news?

mtreinish commented:
Since there hasn't been an update on this in > 1 year and there are some merge conflicts and outstanding review feedback that needs to be addressed I'm going to close this as stale. But, please feel free to reopen this if you have the bandwidth to update it and fix the issues.
